File: array/rotate/48旋转图像.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        # 先沿对角线镜像对称二维矩阵
        for i in range(n):
            for j in range(i, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        # 然后反转二维矩阵的每一行
        for i in range(len(matrix)):
            self.reverse(matrix[i])
        return matrix

    def reverse(self, row):
        i = 0
        j = len(row) - 1
        while j > i:
            row[i], row[j] = row[j], row[i]
            i += 1
            j -= 1

File: array/rotate/48旋转图像.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix[0].size();
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        }
        for (int i = 0 ; i < n; i++) {
            reverse(matrix[i]);
        }
    }

    void reverse(vector<int>& row) {
        int i = 0;
        int j = row.size() - 1;
        while (i < j) {
            int tmp = row[i];
            row[i] = row[j];
            row[j] = tmp;
            i++;
            j--;
        }
    }
};